storage: make chunk_size configurable at runtime

jcsp opened the pull request:
## Cover letter

Where as previously a hardcoded 16kB write size was used, make this size configurable so that it can be tuned for storage devices that are less good at tiny writes, and/or heavily loaded systems where the sheer number of writes in flight can be an issue. 

Fixes: https://github.com/vectorizedio/redpanda/issues/272

## Release notes

A new configuration setting `append_chunk_size` enables tuning the disk write size (default is 16kB, same as what was previously hardcoded).  No change to behavior unless the setting is actively changed by the operator.

jcsp commented:
We could do fewer reads of the config setting if we passed around chunk size as an argument more, but I'm not sure how needed that is -- a shard-local config read isn't particularly expensive.

ztlpn commented on src/v/storage/segment_appender_chunk.h:
```diff
@@ -24,11 +25,10 @@
     explicit segment_appender_chunk(size_t alignment)
       : _alignment(alignment)
-      , _buf(ss::allocate_aligned_buffer<char>(chunk_size, alignment)) {
+      , _chunk_size(config::shard_local_cfg().append_chunk_size())
```
Purely subjectively I feel that for `segment_appender_chunk`s (them being data-like disposable objects without much behavior) chunk size should be passed in constructor by whoever uses them and for other cases it's okay to use the value from the config singleton.

jcsp replied:
I'm inclined to agree -- we construct these in batch in chunk_cache so it's easy to fetch the config once.  Updated.

jcsp commented:
- Rebased on master to fix resolve conflict & get CI fixes
- Made chunk_size a constructor param of segment_appender_chunk instead of fetching config
- Added a validator to cleanly handle non-aligned sizes or sub-4096 sizes

ztlpn commented on src/v/storage/segment_appender_chunk.h:
```diff
@@ -24,11 +25,10 @@
       : _alignment(alignment)
-      , _buf(ss::allocate_aligned_buffer<char>(chunk_size, alignment)) {
+      , _chunk_size(size)
+      , _buf(ss::allocate_aligned_buffer<char>(_chunk_size, alignment)) {
```
Small comment: there is some potential for confusion here as the order in the constructor is different than the order in the seastar function

dotnwat replied:
I guess for any reasonable values the alignment check would fail. but I agree its a little confusing

jcsp replied:
I've swapped these now.

dotnwat commented on src/v/storage/chunk_cache.h:
```diff
@@ -41,6 +41,17 @@ class chunk_cache {
     ~chunk_cache() noexcept = default;
 
+    /** Validator for chunk size configuration setting */
+    static std::optional<ss::sstring> validate_chunk_size(const size_t& value) {
```
👍

dotnwat reviewed (approved):
nice

dotnwat commented:
restarted debug build. probably something transient.

jcsp commented:
Tweaked the argument order to segment_appender_chunk and rebased on latest dev -- this PR should be good to merge

dotnwat commented:
@jcsp just a reminder to (when possible) use `--keep-base` so that intra-diff is small when force pushing https://github.com/vectorizedio/redpanda/compare/ab2f09ff4d243a15f0115cf1e33c410e2e9e3e6c..66b74c7c165ed4903d57ae8432b5f24afe46e2e2
